src/Avalonia.Input/KeyboardDevice.cs

```diff
@@ -211,12 +211,18 @@ public void ProcessRawEvent(RawInputEventArgs e)
                         {
                             var bindings = (currentHandler as IInputElement)?.KeyBindings;
                             if (bindings != null)
-                                foreach (var binding in bindings)
+                            {
+                                // Create a copy of the KeyBindings list.
+                                // If we don't do this the foreach loop will throw an InvalidOperationException when the KeyBindings list is changed.
+                                // This can happen when a new view is loaded which adds its own KeyBindings to the handler.
+                                var cpy = bindings.ToArray();
+                                foreach (var binding in cpy)
```

Comment:
Hmm, this is going to result in an array being copied for every keypress, which I suspect @MarchingCube will have something to say about ;)

Reply:
I wonder if it would be possible to copy the array only when a matching key binding is found?

Reply:
I am not sure if I fully understand what you are saying, but maybe this works?
```csharp
var bindingsToHandle = bindings
    .Where(x => x.Gesture?.Matches(ev) == true)
    .ToArray();

foreach (var binding in bindingsToHandle)
{
    if (ev.Handled)
        break;
    binding.TryHandle(ev);
}
```
This will create an array with all the KeyBindings which should be evaluated. If there are no bindings to be evaluated then the array will be empty, and nothing will happen.
A small issue with this is that the `x.Gesture?.Matches(ev)` is copied from inside `binding.TryHandle(ev)` and therefore this check will run twice. Once here and once inside the `TryHandle` method. Changing this would 'break' the API (though no other code within AvaloniaUI seems to call this method).
I also don't know if using LINQ is acceptable in this context.

Reply:
Yeah something like this should work. Using LINQ here isn't ideal as that itself will allocate but I've pushed a change which does that check manually to your branch, along with a unit test.

Really sorry it took so long to get back to this.